File: eval_test.go
package re

import (
	"fmt"
	"io"
	"strings"
	"testing"
)

type testCase struct {
	name   string
	input  string
	output string
}

var basicCases = []testCase{
	testCase{
		name:   "curry as arg",
		input:  `((lambda (f) (f 42)) (+ 1))`,
		output: "43",
	},

	// testCase{
	// 	name: "let-variable-shadow",
	// 	input: `(do (set (quote f) (lambda (a b)
	// 				   (let a 3 a)) (f 4 5)))`,
	// 	output: "3",
	// },

	// 	testCase{
	// 		name: "let variable shadow",
	// 		input: `(let Result 123
	// (let Result 456
	// 	  (if (= Result 456)
	// 	      true
	// 	      Result)))`,
	// 		output: "true",
	// 	},

	testCase{
		name:   "curry",
		input:  `(do (set (quote f) (lambda (x y z) y)) ((f 1 2) 3))`,
		output: "2",
	},

	testCase{
		name:   "curry-partial",
		input:  `((lambda (x) (lambda (y) (lambda (z) (+ x z)))) 1 2 3)`,
		output: "4",
	},

	// testCase{
	// 	name:   "trap-let",
	// 	input:  "(trap-error (let X 666 42) (lambda E (cons --> (cons A ()))))",
	// 	output: "42",
	// },

	testCase{
		name: "curry1",
		input: `(do (set (quote f) (lambda (x)
			 (do (set (quote ignore) (lambda (z w)
			   (lambda (y)
			      z))) (ignore))))
			(((f 1) 2 3) 4))`,
		output: "2",
	},

	testCase{
		name: "fib10",
		input: `(do (set (quote fib) (lambda (i)
	(if (= i 0)
	    1
	    (if (= i 1)
		1
		(+ (fib (- i 1)) (fib (- i 2)))))))
	(fib 10))`,
		output: "89",
	},

	testCase{
		name: "proper tail call",
		input: `(do (set (quote sum) (lambda (r i)
	  (if (= i 0)
	      r
	      (sum (+ r 1) (- i 1)))))
	(sum 0 5000000))`,
		output: "5000000",
	},

	testCase{
		name:   "do in args",
		input:  `(+ (do 1 (do 2 3)) 4)`,
		output: "7",
	},

	testCase{
		name:   "partial primitive",
		input:  `(+ (+ (+ 1 2) 3) 4)`,
		output: "10",
	},

	testCase{
		name:   "do in tail call",
		input:  `(do (set (quote f) (lambda (x y z) (do 1 (do 2 z)))) (f 1 2 3))`,
		output: "3",
	},

	testCase{
		name:   "closure value",
		input:  "((((lambda (x) (lambda (y) (lambda (z) (+ x z)))) 1) 2) 3)",
		output: "4",
	},

	testCase{
		name:   "basic func call",
		input:  `(do (set (quote id) (lambda (x) x)) (id (do 1 (do 2 42))))`,
		output: "42",
	},

	testCase{
		name:   "identify function",
		input:  `(do (set (quote id) (lambda (x) x)) (id 42))`,
		output: "42",
	},

	testCase{
		name:   "basic set",
		input:  `(do (set (quote x) 42) x)`,
		output: "42",
	},

	testCase{
		name:   "basic if",
		input:  `(if true 1 2)`,
		output: "1",
	},

	testCase{
		name:   "curry lambda",
		input:  `((lambda (x) (lambda (y) (lambda (z) z))) 1 2 3)`,
		output: "3",
	},

	testCase{
		name:   "basic lambda",
		input:  `(((lambda (x y) (lambda (z) y)) 1 2) 3)`,
		output: "2",
	},

	testCase{
		name:   "basic do",
		input:  `(do 1 2)`,
		output: "2",
	},

	testCase{
		name:   "basic primitive",
		input:  `(+ 3 7)`,
		output: "10",
	},

	testCase{
		name:   "constant",
		input:  "42",
		output: "42",
	},

	testCase{
		name:   "partial primitive1",
		input:  `((+ 1) 2)`,
		output: "3",
	},

	testCase{
		name:   "partial primitive2",
		input:  `(((+) 1) 2)`,
		output: "3",
	},
}

func TestBasic(t *testing.T) {
	ctx := New()
	for _, c := range basicCases {
		t.Run(c.name, func(t *testing.T) {
			res := evalString(ctx, c.input)
			if res.String() != c.output {
				fmt.Println("input is:", c.input)
				fmt.Println("output is:", res.String())
				t.Fail()
			}
			if len(ctx.stack) != 0 {
				fmt.Println("unexpected sp after evaluation:", len(ctx.stack))
				t.Fail()
			}
			if ctx.base != 0 {
				fmt.Println("unexpected stack after evaluation:", ctx.base)
				t.Fail()
			}
		})
	}
}

func TestIssue25(t *testing.T) {
	ctx := New()
	evalString(ctx, "(set (quote return) (lambda (x) (lambda (k) (k x))))")
	evalString(ctx, "(set (quote add1) (lambda (n) (return (+ n 1))))")
	res := evalString(ctx, "(add1 4 (lambda (x) x))")
	if res != Integer(5) {
		t.Fail()
	}
}

func evalString(ctx *VM, exp string) Obj {
	r := NewSexpReader(strings.NewReader(exp))
	sexp, err := r.Read()
	if err != nil && err != io.EOF {
		panic(err)
	}

	sexp = macroExpand(ctx, sexp)

	return ctx.Eval(sexp)
}

// func TestTryThrow(t *testing.T) {
// 	ctx := New()
// 	res := evalString(ctx, `(try (lambda (cc handler)
// 		(+ 4 (throw 42 cc handler)))
// 	(lambda (x k)
// 		(k 66)))`)
// 	if res != Integer(70) {
// 		t.Fail()
// 	}
// }

// func TestLoadData(t *testing.T) {
// 	r := NewSexpReader(strings.NewReader("(load \"./init.cora\")"))
// 	sexp, err := r.Read()
// 	if err != nil && err != io.EOF {
// 		panic(err)
// 	}

// 	vm := New()
// 	res := vm.Eval(sexp)
// 	fmt.Println(res.String())
// }

func TestClosureConvert(t *testing.T) {
	// r := NewSexpReader(strings.NewReader(`(lambda (x) x)`))
	// r := NewSexpReader(strings.NewReader(`(lambda (z) (+ x z))`))
	r := NewSexpReader(strings.NewReader(`((((lambda (x) (lambda (y) (lambda (z) (+ x y)))) 1) 2) 3)`))
	sexp, err := r.Read()
	if err != nil && err != io.EOF {
		panic(err)
	}
	exp1, frees := closureConvert(sexp, Nil, Nil, nil)
	fmt.Println("result:", exp1, frees)
}

func TestXXX(t *testing.T) {
	r := NewSexpReader(strings.NewReader(`((+ 1) 2)`))
	// r := NewSexpReader(strings.NewReader(`(((lambda (x y) x) 4) 5)`))
	// r := NewSexpReader(strings.NewReader(`((((lambda (x) (lambda (y) (lambda (z) (+ x z)))) 1) 2) 3)`))
	// r := NewSexpReader(strings.NewReader(`(((lambda (x) (lambda () (+ x 1))) 5))`))
	// r := NewSexpReader(strings.NewReader(`(do (set 'f (lambda (x) x)) (f 44))`))
	// 	r := NewSexpReader(strings.NewReader(`(do
	// (set 'f (lambda (a b c) a))
	// (do
	//  (set 'g (lambda (n) (+ n 1)))
	//  (f 1 (g 5) 4)))`))
	// r := NewSexpReader(strings.NewReader("(load \"../cmd/cora/init.cora\")"))
	// r := NewSexpReader(strings.NewReader(`(do (set (quote sum) (lambda (r i)
	//   (if (= i 0)
	//       r
	//       (sum (+ r 1) (- i 1)))))
	// (sum 0 5000000))`))
	sexp, err := r.Read()
	if err != nil && err != io.EOF {
		panic(err)
	}

	// vm := New()
	// res := vm.Eval(sexp)
	// fmt.Println(res.String())

	// loadFile(vm, "../cmd/cora/init.cora")

	var vm VM
	res := vm.Eval(sexp)
	fmt.Printf("%#v\n", res)

	// var cg CodeGen
	// cg.GenC(code)
}

File: src/eval_test.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "reader.h"
#include <assert.h>
#include "vm.h"

struct testCase {
  char *name;
  char *input;
  char *output;
};

extern void printObj(FILE *to, Obj o);

static void
TestEvalBasic() {
  struct testCase cases[] = {
    {
      "let-variable-shadow",
      "(do (set (quote f) (lambda (a b) (let a 3 a))) (f 4 5))",
      "3",
    },
    {
      "let variable shadow",
      "(let Result 123 \
	(let Result 456 \
		  (if (= Result 456) \
		      true \
		      Result)))",
      "true",
    },
    {
      "curry as arg",
      "((lambda (f) (f 42)) (+ 1))",
      "43",
    },

    {
      "curry",
      "(do (set (quote f) (lambda (x y z) y)) ((f 1 2) 3))",
      "2",
    },

    {
      "curry-partial",
      "((lambda (x) (lambda (y) (lambda (z) (+ x z)))) 1 2 3)",
      "4",
    },
    {
      "curry1",
      "(do (set (quote f) (lambda (x) \
			      (do (set (quote ignore) (lambda (z w) \
						       (lambda (y) \
							z))) (ignore)))) \
		(((f 1) 2 3) 4))",
      "2",
    },

    {
      "fib10",
      "(do (set (quote fib) (lambda (i) \
	(if (= i 0) \
	    1 \
	    (if (= i 1) \
		1 \
		(+ (fib (- i 1)) (fib (- i 2))))))) \
	(fib 10))",
      "89",
    },

    {
      "proper tail call",
      "(do (set (quote sum) (lambda (r i) \
	  (if (= i 0) \
	      r \
	      (sum (+ r 1) (- i 1))))) \
	(sum 0 5000000))",
      "5000000",
    },

    {
      "do in args",
      "(+ (do 1 (do 2 3)) 4)",
      "7",
    },

    {
      "partial primitive",
      "(+ (+ (+ 1 2) 3) 4)",
      "10",
    },

    {
      "do in tail call",
      "(do (set (quote f) (lambda (x y z) (do 1 (do 2 z)))) (f 1 2 3))",
      "3",
    },

    {
      "closure value",
      "((((lambda (x) (lambda (y) (lambda (z) (+ x z)))) 1) 2) 3)",
      "4",
    },

    {
      "basic func call",
      "(do (set (quote id) (lambda (x) x)) (id (do 1 (do 2 42))))",
      "42",
    },

    {
      "identify function",
      "(do (set (quote id) (lambda (x) x)) (id 42))",
      "42",
    },

    {
      "basic set",
      "(do (set (quote x) 42) x)",
      "42",
    },

    {
      "basic if",
      "(if true 1 2)",
      "1",
    },

    {
      "curry lambda",
      "((lambda (x) (lambda (y) (lambda (z) z))) 1 2 3)",
      "3",
    },

    {
      "basic lambda",
      "(((lambda (x y) (lambda (z) y)) 1 2) 3)",
      "2",
    },

    {
      "basic do",
      "(do 1 2)",
      "2",
    },

    {
      "basic primitive",
      "(+ 3 7)",
      "10",
    },

    {
      "constant",
      "42",
      "42",
    },

    {
      "partial primitive1",
      "((+ 1) 2)",
      "3",
    },

    {
      "partial primitive2",
      "(((+) 1) 2)",
      "3",
    },

  };

  struct VM *vm= newVM();
  int pos = 0;
  loadByteCode(vm, pos, cstr("../init.bc"));
  loadByteCode(vm, pos, cstr("../compile.bc"));
  for (int i=0; i<sizeof(cases)/sizeof(struct testCase); i++) {
    struct testCase *c = &cases[i];

    printf("testing case %s ", c->name);

    struct SexpReader r = {.pkgMapping = Nil};
    FILE* f = fmemopen(c->input, strlen(c->input), "r");
    int errCode;
    Obj s = sexpRead(vm, pos, &r, f, &errCode);
    int ref = pos++;
    vmSet(vm, ref, s);
    Obj res = eval(vm, pos, ref);

    char output[512];
    memset(output, 0, 512);
    FILE *to = fmemopen(output, 512, "w");
    printObj(to, res);
    fclose(to);

    int v = strcmp(output, c->output);
    if (v != 0) {
      printf("run test case: %s fail\n", c->name);
      printf("expected: %s\n", c->output);
      printf("actual: %s\n", output);
    }
    assert(v == 0);

    printf("... ok\n");
  }
}

static void
TestTryCatch() {
  struct testCase cases[] = {
    {
      "try-let",
      "(try (lambda () (let X 666 42)) (lambda (E) (cons '--> (cons 'A ()))))",
      "42",
    },

    {
      "no throw",
      "(try (lambda () 42) (lambda (v cc) v))",
      "42",
    },
    {
      "basic throw",
      "(try (lambda () (throw 42)) (lambda (v _) v))",
      "42",
    },
    {
      "resume",
      "(try (lambda () (throw 42)) (lambda (v resume) (resume v)))",
      "42",
    },
    {
      "try in primitive call",
      "(+ (try (lambda () (throw 42)) (lambda (v resume) (resume v))) 1)",
      "43",
    },
    {
      "try in closure call",
      "(do (set 'f (lambda (x) x)) (f (try (lambda () (throw 42)) (lambda (v resume) (resume v)))))",
      "42",
    },
    {
      "throw in primitive call",
      "(try (lambda () (+ (throw 42) 1)) (lambda (v _) v))",
      "42",
    },
    {
      "throw in closure call",
      "(do (set 'f (lambda (a b) b)) (try (lambda () (f (throw 42) 33)) (lambda (v _) v)))",
      "42",
    },
    {
      "resume in primitive call1",
      "(try (lambda () (+ (throw 42) 1)) (lambda (v resume) (resume 66)))",
      "67",
    },
    {
      "resume in primitive call2",
      "(try (lambda () (+ (throw 42) 1)) (lambda (v resume) (resume (+ v 1))))",
      "44",
    },
    {
      "resume in closure call",
      "(do (set 'f (lambda (a b) (+ a b))) (try (lambda () (f (throw 42) 1)) (lambda (v resume) (resume (+ v 1)))))",
      "44",
    },
    {
      "throw in handler",
      "(try (lambda () \
		(try (lambda () \
			(throw \"from body\")) \
		     (lambda (v cc) \
			(throw \"from inner\")))) \
	    (lambda (v1 cc1) \
		v1))",
      "\"from inner\"",
    },
    {
      "try partial curry",
      "((try (lambda () (+ ((throw 4) 7) 5))) (lambda (v cc) (cc (+ v))))",
      "16",
    },
    /* { */
    /*   "curry throw is not supported", */
    /*   "((try (lambda () (throw)) (lambda (v cc) (v))) 666)", */
    /*   "666", */
    /* } */
    {
      "closure ref",
      "((lambda (x y) \
	(try (lambda () \
		(+ x (throw 'ignore))) \
	    (lambda (v resume) (resume y)))) \
	666 1)",
      "667",
    },
    {
      "iterate list",
      "(try (lambda () (map (lambda (x) (throw x)) [1 2 3 4 5])) (lambda (v cc) (cc v)))",
      "(1 2 3 4 5)",
    },
    {
      "throw in deep stack",
      "(do \
 (set 'add (lambda (x y) (+ x y))) \
 (do (set 'f (lambda (n) \
	       (if (= n 0) \
		   (throw 42) \
		 (add (f (- n 1)) 1)))) \
     (try (lambda () \
	    (f 300)) \
      (lambda (v cc) (cc v)))))",
      "342",
    },
    {
      "recover multiple times",
      "(try (lambda () (do (throw 1) (throw 2))) (lambda (v cc) (cc v)))",
      "2",
    },
  };

  struct VM *vm= newVM();
  int pos = 0;
  loadByteCode(vm, pos, cstr("../init.bc"));
  loadByteCode(vm, pos, cstr("../compile.bc"));

  /* char *pkgName = "cora/init"; */
  /* eval(vm, cons(intern("import"), cons(makeString(pkgName, strlen(pkgName)), Nil))); */

  int ref = pos++;
  for (int i=0; i<sizeof(cases)/sizeof(struct testCase); i++) {
    struct testCase *c = &cases[i];

    printf("testing case %s ", c->name);

    struct SexpReader r = {.pkgMapping = Nil};
    FILE* f = fmemopen(c->input, strlen(c->input), "r");
    int errCode;
    Obj s = sexpRead(vm, pos, &r, f, &errCode);
    vmSet(vm, ref, s);
    Obj exp = macroExpand(vm, pos, ref);
    vmSet(vm, ref, exp);
    Obj res = eval(vm, pos, ref);

    char output[512];
    memset(output, 0, 512);
    FILE *to = fmemopen(output, 512, "w");
    printObj(to, res);
    fclose(to);

    int v = strcmp(output, c->output);
    if (v != 0) {
      printf("run test case: %s fail\n", c->name);
      printf("expected: %s\n", c->output);
      printf("actual: %s\n", output);
    }
    assert(v == 0);

    printf("... ok\n");
  }
}

int main() {
  TestEvalBasic();
  TestTryCatch();
}
